src/tools/miri/src/borrow_tracker/stacked_borrows/mod.rs

```diff
@@ -129,7 +128,7 @@ impl NewPermission {
     fn from_box_ty<'tcx>(ty: Ty<'tcx>, kind: RetagKind, cx: &crate::MiriInterpCx<'tcx>) -> Self {
         // `ty` is not the `Box` but the field of the Box with this pointer (due to allocator handling).
         let pointee = ty.builtin_deref(true).unwrap();
-        if pointee.is_unpin(*cx.tcx, cx.param_env()) {
```

Comment:
IMO it'd be better to still have a `param_env()` function, so one can say "the canonical param_env for this interpreter" without having to *know* that in Miri, that is `reveal_all`.

Reply:
the reason I've didn't keep it as a defaulted method in `HasTypingEnv` (previously `HasParamEnv`) is that in a few weeks, I want there to be no functions taking a `ParamEnv` argument inside of the interpreter.Using the `ParamEnv` without also tracking the right `typing_mode` is pretty much never correct. I also change these use-sites to `cx.typing_env().param_env` for now to make this clear, but I don't think that's particularly useful :thinking:

My goal of this refactor is for the interpreter to always either use `cx.typing_env()` or `ty::TypingEnv::fully_monomorphized()` once I am done with this refactor

Reply:
IMO a better intermediate state would be an inherent `fn param_env` on the interpreter. Then when you remove that you can update it to whatever is more appropriate. But if you make it `ParamEnv::releveal_all()` everywhere you won't remember to swap it make to something else later.

> ty::TypingEnv::fully_monomorphized() (if appropriate, i.e. in MIRI and codegen backends) 

That's exactly what I am saying -- I think we shouldn't ever hard-code `fully_monomorphized` in these places, we should follow the same systematic pattern of `cx.typing_env()`. It's shorter and less error-prone. For codegen I have no decision power, but for Miri that's what I'd like to see.

Reply:
that totally makes sense, I agree. We could even add a lint when constructing a `param_env`/`typing_env` in MIRI/codegen, suggesting to instead use the one provided by the `cx`

Reply:
Miri diff looks great, thanks!